pkg/rtc/room.go

```diff
@@ -960,16 +964,26 @@ func (r *Room) pushAndDequeueUpdates(pi *livekit.ParticipantInfo, isImmediate bo
 	shouldSend := isImmediate || pi.IsPublisher
 
 	if existing != nil {
-		if pi.Sid != existing.Sid {
-			// session change, need to send immediately
-			isImmediate = true
-			existing.State = livekit.ParticipantInfo_DISCONNECTED
-			updates = append(updates, existing)
-		} else if pi.Version < existing.Version {
-			// out of order update
-			return nil
-		} else if shouldSend {
-			updates = append(updates, existing)
+		if pi.Sid == existing.Sid {
+			// same participant session
+			if pi.Version < existing.Version {
+				// out of order update
+				return nil
+			} else if shouldSend {
+				updates = append(updates, existing)
+			}
+		} else {
+			// different participant sessions
+			if existing.JoinedAt < pi.JoinedAt {
+				// existing is older, synthesize a DISCONNECT for older and
+				// send immediately along with newer session to signal switch
+				shouldSend = true
+				existing.State = livekit.ParticipantInfo_DISCONNECTED
+				updates = append(updates, existing)
```

Comment:
For this participant, this will be a temporary blip. Basically, that participant would have gotten a `SID` in `JoinResponse`. This means, the participant would get an update with old id (and it will say `DISCONNECTED`) and then new id. JS SDK was applying the update and latching on to old SID and probably ended up in wrong `DISCONNECTED` state too.

This synthesised `DISCONNECT` + new SID after that makes sense for remote participants.

I have not checked the code to verify if this is not sent to participant itself. Should we add code to ensure that this combination is not sent back to the participant and is sent only to the others?

Reply:
You are right. Ideally the old info isn't sent back to the originating participant. We do apply metadata and permission updates from that info.

Reply:
You mean it is applied on server side? Or client also uses it? If client uses it, JS SDK PR from earlier today will bypass that. Cc: @lukasIO

Reply:
Client would apply it. Just saw Lukas' update. That helps on the JS side. Other SDKs will also require that change. I guess it's ok for clients to handle this.

Reply:
Sounds good. Will merge this and have other clients handle this.